rids/iocs/iocs.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2022 Jigsaw Operations LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Combined IOC feed parser interface for the different formats of IOCs."""
+
+
+import urllib.request
+
+from rids.iocs import bad_ip_list
+from rids import rules
+
+
+_PARSERS = {
+  "BAD_IP_LIST": bad_ip_list.FeedParser,
+}
+
+
+def fetch_iocs(rids_config):
+  """Parses the IOC indicated by `ioc_config` into a RuleSet instance.
+  
+  Args:
+    rids_config: { 'iocs': list[dict{'name', 'url', 'format'}] }
+
+  Returns:
+    RuleSet representing all IOCs in the config.
+  """
+  ioc_config = rids_config.get('iocs', None)
+  if not ioc_config:
+    return rules.RuleSet()
+
+  ruleset = rules.RuleSet()
+  for ioc in ioc_config:
+    format = ioc['format']
+    if format not in _PARSERS:
+      raise ValueError('Unrecognized IOC feed format |{format}|')
+
+    parser = _PARSERS[format](ioc_config)
+    with urllib.request.urlopen(ioc_config['url']) as ioc_data:
+      parser.ParseRules(ioc_data, ruleset)
```

Comment:
You can obtain a more [supple design](https://herbertograca.com/2015/12/07/ddd-10-supple-design/) by moving the url fetching into the `Parser` (perhaps rename it). Then the code would read like:
```suggestion
    for rule in _PARSERS[format].CreateRules(ioc_config):
      ruleset.add_rule(rule)
```

More importantly, the parsers can be more powerful. For example, you can have a feed type where  the IOCs are hardcoded in the config, rather than fetched from a URL. Or fetched from a local file.

Consider renaming `format` to `type` or `kind`, meaning the type of the feed spec.

This also has the advantage of making the Parser stateless, which is easier to reason about.

Reply:
I intended to separate out the fetching of IOC data and the parsing of it into Rule instances (which I feel adheres to the Conceptual Contours and Standalone Classes of the blog post you referenced).  It has also made it easier to test the parsing and rule construction because unit tests can provide an io.BytesIO() and not need to fake any calls to urlfetch().

For the extensions you mention -- anything that can be defined in the config can be used when the parser is constructed (the first line, that you deleted).  The urllib..urlopen() call already can read from a file path (using file://...).  I can change the abstraction so that the content gets fetched by the IocSource (ne'e Parser), and yeah a better name is probably needed if it's taking on more responsibility.

Since we're discussing the design of this particular part, I'd like to draw your attention to the need to periodically update these feeds.  Many get new content ~daily and we would want to re-fetch at least that often (ET says not to do it more than hourly, rates may vary).  I think this object here is a good place for that, and rather than being called parser, can be an IocSource which maintains the state of when it should be re-fetched.  It would know how to parse the format too, of course, and it can still separate the Parser into its own class and be kept stateless and easy to reason about, but even iocs.fetch_iocs(...) wouldn't need to know about parsing (though it would still benefit from dispatching on 'format' type).  Periodic re-fetch will be a separate PR.

The `format` here is the data format of the IOC contents.. there isn't a 1:1 mapping with IOC type or kind (in terms of IP-type rules or TLS-type rules).  There are multiple formats available for getting IP-type rules.  It's actually n:n because some formats could produce multiple types of rules.  I think type and kind are less informative than format when it's meant to indicate the data format of the IOC file and how it gets parsed.

I've updated the pydoc in a few places to hopefully make it clearer what is meant by format.